DDD-12.md

```diff
@@ -0,0 +1,326 @@
+
+```java
+public record Order(long id, String name, int price) {}
+```
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class OrderListener {  
+  
+    @DebeziumListener("order")  
+    public void listener(InsertEvent<String, Order> event) {  
+        /// some logic to apply
+    }  
+}
+```
+
+using something similar for the [quarkus kafka library](https://quarkus.io/guides/kafka#jackson-serialization):
+
+```java
+package com.acme.order.jackson;
+
+import io.quarkus.debezium.client.serialization.ObjectMapperDeserializer;
+
+public class OrderDeserializer extends ObjectMapperDeserializer<Order> {
+    public OrderDeserializer() {
+        super(Order.class);
+    }
+}
+```
+
+```properties
+quarkus.debezium.deserializer=com.acme.order.jackson.OrderDeserializer
+```
+
+### Quarkus Debezium SchemaChange Listener
+Debezium automatically detects and captures schema changes in the source database, such as adding or removing columns, modifying data types, or altering primary keys. These changes are parsed from the database's DDL statements and used to update Debezium's internal schema history, ensuring that change events reflect the current table structure. The Quarkus extension can expose a listener to such kind of event:
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class SchemaChangeListener {  
+  
+    @DebeziumSchemaChangeListener()  
+    public void listener(SchemaChangeEvent event) {  
+        /// some logic to apply
+    }  
+}
+```
+
+### Quarkus Debezium Snapshot Listener
+The snapshot is Debezium’s initial data capture phase. It reads the current state of selected tables and emits each row as a change event. This happens once at startup (unless disabled) to ensure consumers start with a complete dataset before live change streaming begins. It can be observed with the Quarkus Extension thanks to the following annotations:
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class SnapshotListener {  
+  
+    @DebeziumSnapshotStarted()  
+    public void started(SnapshotEvent event) {
+        /// some logic to apply
+    }
+    
+    @DebeziumSnapshotCompleted()
+    public void completed(SnapshotEvent event) {
+        /// some logic to apply
+    }
+}
+```
+
+### Quarkus Debezium PostProcessor
+
+In Debezium Embedded Engine, a PostProcessor is an internal hook used to modify or transform events right before they are emitted to the application’s `.notifying()` callback. In the quarkus extension can be accessible by:
+
+```java
+import io.debezium.engine.InsertEvent;
+import jakarta.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped  
+class CustomPostProcessor {  
+  
+    @DebeziumPostProcessor()  
+    public void process(Object key, Struct value) {
```

Comment:
There should be a return type here. Isn't it?

Reply:
Well actually the interface is written in this [way](https://github.com/debezium/debezium/blob/6a37d34f627f878280f692d51e498c1c64840605/debezium-core/src/main/java/io/debezium/processors/spi/PostProcessor.java#L30).

and the comment:

>  * While this may seem similar to a Kafka Transformation, the difference is that a post processor
>* operates on the raw {@link Struct} objects which are still mutable, allowing for a variety of
>  * use cases that wish to have the full data set of a row for the post-processing step.

IMHO, it's really powerful, and I'm not sure I’d want to give that much power to a *quarkus developer*. On the other hand, I’m not sure how to express it differently

Reply:
Do you have any suggestion?

Reply:
@Naros Any particular reason why this was not something like `public Struct process(Object key, Struct value)` so letting the implementors to provide the post processed `Struct`?

Reply:
@mfvitale the main reason was to enforce the notion that you shouldn't change the value `Struct` schema layout, that you should only be mutating the values related to fields within the `Struct`. In some places where the `EventDispatcher` calls into the `PostProcessor`, the envelope representation has already been computed before prior calls in the stack.  Changing the `Struct` layout in a `PostProcessor` would require a recreation of the envelope.

Reply:
I am not sure that we want to be in the habit of where we need to re-create the envelope schema anytime a user uses a `PostProcessor`, as the entire point of the `TableSchemaBuilder#create` method was to do this once everytime a table is added or altered.

Reply:
I'd keep the things as they are now - so no change in schema. We can think about opening this option in future but for now I'd like to have the extension provide the same feature set we have in Debezium core. Let's avoid scope creep.

Reply:
@Naros sounds good. Let's keep it as is.